Surface arrangement memory size/capacity

antiguru opened the pull request:
Measure the size of arrangements in memory.

This PR adds support to measure the size of arrangements:
* Total allocated capacity,
* used capacity,
* number of allocations.

### Motivation

This PR adds a known-desirable feature: #19348

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

antiguru commented:
This should be ready for review. For some arrangements, it reports an under-approximation of allocated memory because we treat the diff field not in a special way, which misses allocations it might reference. This is the case for some of the reductions.

antiguru commented:
This PR is approaching a state where we can discuss remaining work. I think at least the following should be included in here:
* Currently, only a single `_raw` collection is exposed. We'd need to add convenience views on top of this to aggregate across workers and dataflows.
* The introspection source we expose doesn't have the count in the diff field, but materializes it automatically in the actual row. This is bad because it means we have to do more work than if we could simply maintain the count. However, this is not possible at the moment, because we're counting used size, raw capacity, and number of allocations at the same time. To move the values to the diff field, we'd need three separate sources.
* We're printing warnings when the logging operator cannot be activated from dataflow logging. This is helpful to determine which arrangements still need to be logged, but doesn't make much sense for ad-hoc queries because the activator will most likely not be available yet/anymore when we process differential logs. I don't know a smart approach that would allow us to keep the warning and don't produce bogus warnings for short-lived dataflows.
* We're moving several operators and abstractions into `mz-compute`. This feels restrictive, and we might want to consider splitting the logging infrastructure out of this, together with the abstraction introduced here. This will become more relevant once we actually work on cluster unification (#17413).
* We're undercounting the size of `DataflowError` because we only consider the immediate allocation, but not the transitive closure of reachable allocations.. Maybe we could implement `Columnation` for it and get accurate accounting? It seems hard because it's a recursive data structure...
* I have not looked into how many resources this feature requires. It adds a significant number of operators, activates these operators regularly, and maintains the introspection data. Each can be expensive.
  * We could use a rate-limiting activator if activations turn out to be problematic.
* The name `mz_arrangement_size_raw` isn't great, because there is also `mz_arrangement_sizes...`. We need to come up with something better.
* We should adjust the memory visualizer to print both number of records as well as memory requirements for operators.

def- commented on src/compute/src/render/threshold.rs:
```diff
@@ -96,21 +96,29 @@ where
-    let negatives = match &arrangement {
-        ArrangementFlavor::Local(oks, _) => {
-            threshold_arrangement(oks, "Threshold retractions local", |count| *count < 0)
+    let (oks, negatives, errs) = match &arrangement {
```
This entire block also appears untested and quite interesting (this actually seems suspicious to me, if there should be a test for this I can check if coverage is broken in some way)

antiguru replied:
Yep, the whole function cannot be exercised through SQL because the optimizer never generates a plan for it. Although it'd be nice to actually use it, it seems we're better off removing it, and bringing it back once we actually can use it. I believe the implementation has a bug that this change fixes.

antiguru replied:
Dead code should have been removed.

antiguru commented on src/compute/src/extensions/collection.rs:
```diff
@@ -0,0 +1,73 @@
+        Arranged<G, TraceAgent<Tr>>: ArrangementSize,
+    {
+        if must_consolidate {
+            self.mz_consolidate::<Tr>(name)
```
Interesting, I'll defer to @vmarcos who might have more context on why this is! It might be that it only triggers within ad-hoc selects with monotonic optimizations enabled, and this might not be the case within tests.

vmarcos replied:
Yes, the line should only be executed if the feature flag `enable_monotonic_oneshot_selects` is `on` and we have a one-shot select where the optimization applies, e.g., a query with `MAX`/`MIN` or a top-k pattern. AFAIK, the tests are currently run with the feature flag turned `off`, so that we get to test our regular optimization path. The design doc #19250 has some suggestions on how to address this issue, but no solution is implemented yet at this very moment. Perhaps it would make sense to run coverage with the feature flag `on` as a validation?

antiguru commented on src/compute/src/logging/compute.rs:
```diff
@@ -586,6 +681,80 @@ impl DemuxHandler<'_, '_> {
+        let Some(state) = self.state.arrangement_size.get_mut(&operator_id) else {return};
+
+        if state.size == size && state.capacity == capacity && state.count == count {
+            return;
```
Yep, it's checked elsewhere, so we shouldn't see this happening. Thanks for pointing it out; I'll remove it.

antiguru commented on src/compute/src/extensions/operator.rs:
```diff
@@ -0,0 +1,344 @@
+        self.arrange_named(name).log_arrangement_size()
+    }
+
+    fn mz_arrange_core<P, Tr>(&self, pact: P, name: &str) -> Arranged<G, TraceAgent<Tr>>
```
Interesting. At the moment it's only used by the introspection code, could it be that introspection is turned off for coverage runs?

def- replied:
How to turn on/off introspection? Coverage is run the same way as normal tests, so this would also mean we are not using it in CI tests probably.

teskje replied:
It should be switched on in CI. There are both testdrive and sqllogictests that query introspection sources.

umanwizard commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -324,11 +324,14 @@ The `mz_arrangement_sharing` view describes how many times each [arrangement] in
+| `batches`     | [`bigint`] | The number of batches in the arrangement.                                                                                 |
+| `size`        | [`bigint`] | The utilized size in bytes of the arrangement.                                                                            |
+| `capacity`    | [`bigint`] | The capacity in bytes of the arrangement. Can be larger than the size.                                                    |
+| `count`       | [`bigint`] | The number of allocations backing the arrangement.                                                                        |
```
`count` is a confusing name IMO (too generic).

antiguru replied:
Great point! Changed to `allocations`.

umanwizard commented on src/compute/src/extensions/operator.rs:
```diff
@@ -0,0 +1,368 @@
+        move |frontiers| {
+            input.for_each(|_time, _data| {});
+            if frontiers[0].is_empty() {
+                // TODO(mh): Correctness
+                // This is not correct: We can maintain a trace after the arrangement operator
+                // has been dropped, which means that we report that the trace does not consume
+                // space, but it might still do so. Fixing it would require Differential logging
+                // to signal when trace sharing drops to 0 (or 1 because this operator shares one
+                // copy.)
```
@frankmcsherry  had some ideas for how to deal with this issue, but maybe it doesn't matter.

antiguru replied:
I changed it such that it'll retract the information once the trace isn't shared anymore.

jkosh44 reviewed (approved):
Adapter changes LGTM (I did not review the `.jsx` files).

jseldess reviewed (approved):
Docs LGTM. 

Based on the linked issue, it seems these changes will enable users to understanding their most expensive or memory-intensive indexes and materialized views. Is that right? If so, would it make sense to revise or complement [this troubleshooting FAQ](https://materialize.com/docs/manage/troubleshooting/#why-is-materialize-using-so-much-memory) as part of this PR? Or should this rather be a follow-up docs issue?

antiguru commented:
> Based on the linked issue, it seems these changes will enable users to understanding their most expensive or memory-intensive indexes and materialized views. Is that right? If so, would it make sense to revise or complement [this troubleshooting FAQ](https://materialize.com/docs/manage/troubleshooting/#why-is-materialize-using-so-much-memory) as part of this PR? Or should this rather be a follow-up docs issue?

I'd prefer to do this in a follow-up issue. I filed #19500 and #19501 to not forget about it!

antiguru commented on src/compute/src/extensions/operator.rs:
```diff
@@ -0,0 +1,371 @@
+    G::Timestamp: Lattice + Ord,
+    K: Data + Columnation,
+    T: Lattice + Timestamp,
+    R: Semigroup,
```
This is somewhat incorrect: `R` can be a complex type, such as a reduction monoid, but we're not counting this correctly without a `Columnation` bound. I'm working on a fix.

antiguru replied:
It turns out that Differential doesn't currently support region-allocated `R`. This is because `OrderedLeaf` has a parameter to specify the container type, but where it's used in `OrdValBatch`, it's not specified. This requires a change to Differential to fix, or us defining our own batch types.

antiguru replied:
I looked into this a bit more. Differential keeps for every leaf a `Vec<(T, R)>` maintaining how often a value appears at a certain time. When advancing time, it does so by advancing the `T` in-place, which requires mutable access to the data. With `Vec`, this is safe, but switching to `TimelyStack` does not allow safe mutable access to its elements. Indeed, it would only be safe if `advance_by` would not cause a reallocation, which in general we cannot guarantee. This means we cannot have a bound of `Columnation` on `R`, making it impossible to region-allocate various monoids.

For details, see TimelyDataflow/differential-dataflow#391.

teskje commented on src/storage-client/src/types/errors.rs:
```diff
@@ -411,6 +411,15 @@ pub enum DataflowError {
+    use timely::container::columnation::{CloneRegion, Columnation};
+
+    impl Columnation for DataflowError {
+        type InnerRegion = CloneRegion<DataflowError>;
```
`DataflowError` does contain pointers. Why is it okay to use a `CloneRegion` for it?

antiguru replied:
I'll add a justification, but the TL;DR is that it's correct, it's complicated to implement columnation (recursive type, wide enums), and we accept the performance penalty, which should be on par to what we currently have with storing the errors in vectors. For reference, here is an incomplete attempt to implement `Columnation` for `DataflowError`: https://gist.github.com/antiguru/2608b8bb40f1a948aab0761a4ac9d0e5

teskje commented on test/testdrive/introspection-sources.td:
```diff
@@ -282,15 +282,15 @@ mv1_primary_idx mv1
-> SELECT records, batches FROM mz_internal.mz_dataflow_arrangement_sizes WHERE name='ii_empty'
-0 0
+> SELECT records, batches, size, capacity, allocations FROM mz_internal.mz_dataflow_arrangement_sizes WHERE name='ii_empty'
+0 0 0 0 0
```
Would it be possible to also test that after dropping an arrangement, it entries disappear from `mz_arrangement_sizes` and/or related collections? I'd like to have something that ensures that we correctly retract the three new values on arrangement drops.

antiguru replied:
Test added.

teskje reviewed (changes requested):
This looks mostly good, but there are two things I'd like to ensure:

1. That we have user docs for all new relation fields.
2. That the retraction of arrangement size logging works correctly.

My other comments are nits and/or questions.

antiguru commented on src/compute/src/extensions/mod.rs:
```diff
@@ -0,0 +1,16 @@
+
+pub(crate) mod collection;
+pub(crate) mod operator;
+pub(crate) mod reduce;
```
Moved all reductions into `reduce`, renamed `operator` to `arrange`.

antiguru commented on src/compute/src/logging/compute.rs:
```diff
@@ -77,6 +80,39 @@ pub enum ComputeEvent {
+        operator: usize,
+        /// Number of distinct heap allocations backing the arrangement.
+        allocations: isize,
+    },
```
Yup, thanks, fixed!

antiguru commented on src/compute/src/logging/compute.rs:
```diff
@@ -296,14 +375,29 @@ pub(super) fn construct<A: Allocate>(
-#[derive(Default)]
-struct DemuxState {
+struct DemuxState<A: Allocate> {
+    /// The scope of this operator.
```
Thanks!

antiguru commented on src/compute/src/logging/compute.rs:
```diff
@@ -586,6 +713,82 @@ impl DemuxHandler<'_, '_> {
+            .arrangement_heap_size
+            .give((datum, ts, Diff::cast_from(size)));
+
+        state.size = size;
```
Definitely not correct! Should be better now, but a test is still missing. Will add one.

antiguru replied:
Test added.

antiguru commented on src/compute/src/extensions/mod.rs:
```diff
@@ -0,0 +1,16 @@
+//! Rendering of `MirRelationExpr::Join` operators, and supporting types.
+//!
+//! Consult [mz_compute_client::plan::join::JoinPlan] documentation for details.
+
```
Done.

antiguru commented on src/compute/src/logging/compute.rs:
```diff
@@ -301,48 +301,26 @@ pub(super) fn construct<A: Allocate + 'static>(
+            (PeekDuration, peek_duration),
+            (ArrangementHeapSize, arrangement_heap_size),
+            (ArrangementHeapCapacity, arrangement_heap_capacity),
+            (ArrangementHeapAllocations, arrangement_heap_allocations),
```
Done!.

antiguru commented on src/compute/src/extensions/operator.rs:
```diff
@@ -94,44 +119,57 @@ where
 }
 
-impl<G, R> MzArrange<G, DataflowError, (), R> for Collection<G, DataflowError, R>
+pub struct KeyCollection<G: Scope, D, R: Semigroup = usize>(Collection<G, D, R>);
```
Yep, done.

antiguru commented on src/compute/src/extensions/operator.rs:
```diff
@@ -94,44 +119,57 @@ where
+pub trait IntoKeyCollection {
+    type Output;
+    fn into_key_collection(self) -> Self::Output;
+}
```
Done. It means more type annotations, but actually reduces the lines of code because of better formatting.

vmarcos reviewed (commented):
This is very exciting, and will definitely surface some very valuable information! I have a few comments regarding documentation, some general questions, and one point for discussion in the change to basic aggregates in `reduce.rs`.

antiguru commented on src/compute/src/extensions/reduce.rs:
```diff
@@ -75,9 +79,50 @@ where
+    {
+        // Allow access to `reduce_core` since we're within Mz's wrapper.
+        #[allow(clippy::disallowed_methods)]
+        self.reduce_core::<_, T2>(name, move |key, input, output, change| {
+            if !input.is_empty() {
+                logic(key, input, change);
+            }
+            change.extend(output.drain(..).map(|(x, d)| (x, d.negate())));
+        })
```
Good question! It's probably better to call `reduce_abelian`. Removing the `consolidate` call eliminates redundant work, because the reduction thinker immediately consolidates the changes after the closure returns. We should fix this in Differential!

vmarcos replied:
Ah, that makes sense. Agreed that it's better to call `reduce_abelian` here and open a PR for DD!

antiguru commented on src/compute/src/render/reduce.rs:
```diff
@@ -612,16 +607,16 @@ where
         input
-            .arrange_named::<RowSpine<(Row, Row), _, _, _>>("Arranged ReduceInaccumulable")
-            .reduce_abelian::<_, RowSpine<_, _, _, _>>(
+            .mz_arrange::<RowSpine<Row, Row, _, _>>("Arranged ReduceInaccumulable")
```
I reverted it to use the current behavior. The `KeyCollection` type makes this more explicit than what we currently have to rely on.

antiguru commented on src/compute/src/logging/differential.rs:
```diff
@@ -119,28 +125,64 @@ pub(super) fn construct<A: Allocate>(
+            .log_register()
+            .get::<ComputeEvent>("materialize/compute");
+        if let Some(logger) = compute_logger {
+            // Arrange the sharing information and reduce it. This will reveal when an arrangement
```
Fixed!

antiguru commented on src/compute/src/extensions/arrange.rs:
```diff
@@ -0,0 +1,317 @@
+                batch.layer.vals.keys.heap_size(&mut callback);
+                vec_size(&batch.layer.offs, &mut callback);
+                vec_size(&batch.layer.vals.offs, &mut callback);
+                vec_size(&batch.layer.vals.vals.vals, &mut callback);
```
Yep, that's correct. We either need to specialize for different `R` values and do a heap traversal to count the allocations correctly, or we indeed interpret this as a lower bound. It's not great, but I'm not excited about actually traversing all pointers to determine the heap size. Columnation has the nice property that we only need to look into the allocator to determine heap stats, which should be significantly cheaper than actually traversing the object graph.

antiguru commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -2919,7 +2981,10 @@ pub const MZ_DATAFLOW_ARRANGEMENT_SIZES: BuiltinView = BuiltinView {
             mo.name,
             COALESCE(sum(mas.records), 0) AS records,
-            COALESCE(sum(mas.batches), 0) AS batches
+            COALESCE(sum(mas.batches), 0) AS batches,
+            COALESCE(sum(mas.size), 0) AS size,
+            COALESCE(sum(mas.capacity), 0) AS capacity,
+            COALESCE(sum(mas.allocations), 0) AS allocations
```
Updated the documentation.

antiguru commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -2231,10 +2249,14 @@ GROUP BY
+    pg_catalog.SUM(records) as records,
+    pg_catalog.SUM(size) as size,
+    pg_catalog.SUM(capacity) as capacity,
+    pg_catalog.SUM(allocations) as allocations
```
I'd like to leave the renaming to a future PR and not make it part of this change.

antiguru replied:
Filed #19562 to not forget about this!

vmarcos reviewed (approved):
The updates addressed all of my review points, thanks! So no more requests from my side other than the tests that Jan suggested adding.

antiguru commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -498,22 +507,43 @@ The `mz_peek_durations_histogram` view describes a histogram of the duration in
+| `records`     | [`numeric`] | The number of records in the operator.                                                       |
+| `size`        | [`bigint`]  | The utilized size in bytes of the arrangement.                                               |
+| `capacity`    | [`bigint`]  | The capacity in bytes of the arrangement. Can be larger than the size.                       |
+| `allocations` | [`bigint`]  | The number of separate memory allocations backing the arrangement.                           |
```
Updated the documentation.

antiguru commented:
I think I addressed all comments, so if you have time, please take another look. I also kicked off a [nightly run](https://buildkite.com/materialize/nightlies/builds/2500) and [coverage](https://buildkite.com/materialize/coverage/builds/123).

I had to change the logic that uses arrangements with reductions to determine when an arrangement trace wasn't shared anymore because it fails to work when the arrangement is created and dropped within a single log window. In this case, the arrangement will not reveal any information about it because the events cancel each other out. We might want to think if there's a way we could still achieve the same behavior. For the time being, I switched to a map-based implementation.

def- commented:
> I also kicked off a [nightly run](https://buildkite.com/materialize/nightlies/builds/2500) and [coverage](https://buildkite.com/materialize/coverage/builds/123).

The SQLsmith failures are unrelated. The feature benchmark regression is probably expected with this change?

teskje commented on test/testdrive/introspection-sources.td:
```diff
@@ -282,15 +282,38 @@ mv1_primary_idx mv1
+> SELECT records, batches, size, capacity, allocations FROM mz_internal.mz_dataflow_arrangement_sizes WHERE name='ii_empty'
+0 0 0 0 0
+
+# Tests that arrangement sizes are approximate
```
Is this comment missing some words?

teskje commented on src/compute/src/logging/differential.rs:
```diff
@@ -305,6 +282,17 @@ impl DemuxHandler<'_, '_, '_> {
+                self.state.sharing.remove(&op);
+                logger.log(ComputeEvent::ArrangementHeapSizeOperatorDrop { operator: op });
+            }
+        }
```
This works because `compute_logger` will always be `Some` in practice, but if we ever change the code to make it temporarily `None` the `sharing` tracking might become inconsistent. How about we always update `sharing` and only gate the actual `log` call behind `if let Some(logger)`?

antiguru replied:
Yep, we only track the sharing information if the logger is `Some`. That's currently the only case when the information is needed. Do you think we should always track this information?

teskje replied:
It would be more future-proof that way, but I'm fine with leaving it as is.

Out of scope for this PR, but I wonder if we could get rid of the `if let Some(logger)` pattern throughout the compute code and make the logger non-optional instead. I _think_ we always have a compute logger, except perhaps during initialization.

teskje reviewed (approved):
LGTM, thanks!

antiguru commented:
Verified locally that the `GroupBy` regression goes away with the last commit.
